File: libpkgupd/builders/autoconf.cc
#include "autoconf.hh"

#include "../exec.hh"
#include "../recipe.hh"

namespace rlxos::libpkgupd {
bool AutoConf::compile(Recipe const& recipe, std::string dir,
                       std::string destdir,
                       std::vector<std::string> const& environ) {
  auto configure = recipe.configure();

  bool same_dir = false;
  if (recipe.node()["autoconf-dir"]) {
    same_dir = recipe.node()["autoconf-dir"].as<bool>();
  }

  std::string configurator = "./configure";
  if (!same_dir) {
    configurator = "../configure";
    dir += "/_pkgupd_builddir";
  }

  if (configure.find("--prefix") != std::string::npos) {
    configure = " --prefix=" + PREFIX + " --sysconfdir=" + SYSCONF_DIR +
                " --libdir=" + LIBDIR + " --libexecdir=" + LIBEXEDIR +
                " --bindir=" + BINDIR + " --sbindir=" + SBINDIR +
                " --datadir=" + DATADIR + " --localstatedir=" + CACHEDIR + " " +
                configure;
  }
  if (int status =
          Executor().execute(configurator + " " + configure, dir, environ);
      status != 0) {
    p_Error = "failed to do meson configuration";
    return false;
  }

  if (int status = Executor().execute("make " + recipe.compile(), dir, environ);
      status != 0) {
    p_Error = "failed to build with meson";
    return false;
  }

  auto install = recipe.install();
  if (install.find("install ") == std::string::npos) {
    install = "install " + install;
  }

  if (int status = Executor().execute("make DESTDIR=" + destdir + " " + install,
                                      dir, environ);
      status != 0) {
    p_Error = "failed to install with meson";
    return false;
  }

  return true;
}
}  // namespace rlxos::libpkgupd

File: builder/buildsystem/autoconf/autoconf.go
package autoconf

import (
	"fmt"
	"os"
	"os/exec"
	"path"
	"runtime"
	"strings"

	"github.com/itsmanjeet/pkgupd/recipe"
	"github.com/itsmanjeet/pkgupd/utils"
)

type AutoConf struct {
}

func (a AutoConf) Compile(mod *recipe.Module, srcdir string, destdir string) error {
	execute := func(bin string, args ...string) error {
		for i := range args {
			args[i] = os.ExpandEnv(args[i])
		}
		cmd := exec.Command(bin, args...)
		cmd.Dir = path.Join(srcdir, mod.Dir)
		if mod.BuildDir {
			cmd.Dir = path.Join(cmd.Dir, utils.BUILD_DIR)
		}
		cmd.Stderr = os.Stdout
		cmd.Stdin = os.Stdin
		cmd.Stdout = os.Stdout
		cmd.Env = os.Environ()
		cmd.Env = append(cmd.Env, "pkgupd_srcdir="+cmd.Dir)
		return cmd.Run()
	}

	if err := os.MkdirAll(path.Join(srcdir, mod.Dir, utils.BUILD_DIR), 0755); err != nil {
		return fmt.Errorf("failed to create build directory %s", err)
	}

	convert := func(data string, def []string, hotword string) []string {
		if len(data) == 0 {
			return def
		}
		userArgs := strings.Split(strings.Trim(data, " \n"), " ")
		if strings.Contains(data, hotword) {
			return userArgs
		}

		def = append(def, userArgs...)
		return def
	}

	configure := "./configure"
	if mod.BuildDir {
		configure = "../configure"
	}

	if err := execute(configure, convert(mod.Configure, []string{"--prefix=/usr", "--sysconfdir=/etc", "--libdir=/usr/lib", "--libexecdir=/usr/lib", "--sbindir=/usr/bin", "--bindir=/usr/bin", "--datadir=/usr/share", "--localstatedir=/var"}, "--prefix")...); err != nil {
		return fmt.Errorf("failed to configure %s", err)
	}

	if err := execute("make", convert(mod.Compile, []string{fmt.Sprintf("-j%d", runtime.NumCPU())}, "-C")...); err != nil {
		return fmt.Errorf("failed to compile %s", err)
	}

	if err := execute("make", convert(mod.Compile, []string{"install", "DESTDIR=" + os.Getenv("DESTDIR")}, "install")...); err != nil {
		return fmt.Errorf("failed to compile %s", err)
	}

	return nil
}
